1단계 - 예약 기능 구현

OikimiO opened the pull request:
안녕하세요ㅎ

이번 사전 학습을 진행하는 목표 중 하나로 테스트 코드로 문서화를 진행했습니다.
그래서 완벽하지 않은 상태에서 제 개인적인 기준..^^ BDD 관점의 테스트 코드를 만들어 보려 노력했습니다. 
그런데 테스트를 진행하며 각각의 테스트를 진행하면서는 문제 없었던 것들이.. 통합테스트를 하며 문제가 발생했습니다.

문제가 발생한 원인은 테스트 코드간의 의존성이 생겨.. 데이터를 저장하지 않아야할 사항에 저장이 된 상태로 테스트가 진행되거나 하는.. 문제가 
있었습니다..;;ㅎ
제가 알기로도 테스트 코드간의 의존성은 생겨선 안되고 독립적으로 구성이되어야 하는데.. 현재는 그렇지 못하다는 한계가 존재합니다..ㅎ

테스트 코드가 미완성인 상태로 코드를 올립니다.. 혹시나 코드를 보시고 테스트의 방향성이나 고쳐야 할 부분들이 있다면 알려주시면 감사하겠습니다.

OikimiO committed:
feat: index.html 출력을 위한 Controller 추가

OikimiO committed:
feat: 에약 관리 API1-1. Controller 메서드 추가 외 객체 추가

- 조회/저장/삭제를 처리할 메서드 추가
- 예약을 관리할 엔티티 클래스 추가
- 요청/응답 클래스 Record로 구현

OikimiO committed:
feat: 예약 관리 API1-1. 예약 관리를 위한 Entity&저장 DTO 추가

- 외부의 의존도를 낮추기 위해 입력/조회/수정 인터페이스 구성
- 조회 후 데이터를 수정할 없도록 구성
- Reservation을 관리하는 List는 static하게 구성
- Reservation 저장을 위한 DTO 내용 구성

OikimiO committed:
feat: 예약 관리 API1-1. 예약 관리 로직 처리를 위한 Service&조회 DTO 추가

- 조회를 위한 임시 데이터 입력 메서드 추가
- 조회/저장/삭제시 Reservation의 인터페이스로만 호출(결합도 낮아짐)
- 조회시 곧바로 Entity가 아닌 DTO를 거쳐 응답

OikimiO committed:
refactor: 예약 관리 API1-1. MissionStep test에 맞춰 각 객체 수정

- AdminService: Reservation 임시 데이터 및 초기화 기능 제거
- Reservation: 초기화 기능 제거
- ReadReservationResponse: Entity 단건 변환을 위해 생성자 public으로 변경

OikimiO committed:
feat: 예약 관리 API1-2. schema&properties 수정

OikimiO committed:
refactor: 예약 관리 API1-2. AtomicLong&List제거 및 Repository 추가

- Reservation: 변수 AtomicLong&List 제거, 조회/삭제 메서드 제거, 저장 메서드 수정
- AdminServie: Reservation 도메인 로직 AdminRepository가 처리할 수 있도록 코드 수정
- AdminRepository: AdminService의 조회/저장/삭제 메서드 추가

OikimiO committed:
refactor: 예약 관리 1-2. 조회/삭제/저장 기능 구현을 위해 기능 추가&변경

- AdminRepository: JdbcTemplate을 통한 조회/삭제/저장 기능 추가
- AdminService: 저장 후 id값을 return하도록 변경
- ReadReservationResponse: 저장 후 조회를 위한 id, Reservation을 파라미터로 하는 생성자 추가
- Reservation: 조회 후 ResultSet 결과를 입력을 위한 생성자 추가
- SaveReservationRequest: 수정 필요 없어 이전 기능으로 롤백
- schema.sql: reservation 테이블이 있을 경우 drop ddl 추가

OikimiO committed:
feat: 예약 관리 1-3. time 페이지 추가

OikimiO committed:
refactor: 예약 관리 1-3. schema reservation_time 추가 및 reservation 테이블 변경

- schema: reservation time_id 외래키 추가

OikimiO committed:
feat: 예약 관리 1-3. Controller에 시간 관리 메서드 및 DTO 추가

- 조회/저장 DTO 추가

OikimiO committed:
feat: 예약 관리 1-3. Service에 시간 관리 메서드 및 Entity 추가

- AdminService/AdminRepository: 조회/삭제/저장 메서드 추가
- ReadReservationTimeResponse: 도메인 > List 변경 메서드 및 도메인 > DTO 변경 기능 추가
- SaveReservationTimeRequest: 시작시간 파라미터 추가
- ReservationTime: Entity class 추가 후 조회/저장을 위한 메서드 추가

OikimiO committed:
feat: 예약 관리 1-3. Repository 시간 관리 메서드 구현

AdminRepository: 조회/삭제/저장 관련 쿼리 기능 구현
ReadReservationTimeResponse: 조회 응답을 위한 DTO 기능 구현

OikimiO committed:
refactor: 예약 관리 1-3. 기존 Reservation 기능 수정 진행

- ReservationQuery,ReservationTimeQuery: 쿼리문을 클래스로 관리
- ReadReservationResponse: Entity > List & Entity > DTO 기능으로 변경, 생성자 private 처리
- SaveReservationRequest: time > timeId로 변경
- Reservation: ReservationTime 필드 추가 및 read 메서드 변경
- AdminService: Reservation 저장 후 조회 기능 변경
- AdminRepository: Reservation 저장 후 조회 쿼리 기능 추가

OikimiO committed:
refactor: 예약 관리 1-3. 패키지 구조 변경

- controller, dto, entity, repository, service로 패키지 구분

OikimiO committed:
test: Controller  테스트 코드 추가 & 그래들 기능 추가

- AdminControllerTest: 테스트 기능 추가

xrabcde commented on src/main/java/roomescape/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,78 @@
+public class AdminController {
+
+    private static final String BASE_PATH = "/admin";
+    @Autowired
+    private AdminService adminService;
```
의존성을 필드로 주입해주고 계시네요! 
이 방법은 어떤 식으로 동작하며 어떤 장단점이 있을까요? 

[Spring의 다양한 의존성 주입방법](https://tecoble.techcourse.co.kr/post/2020-07-18-di-constuctor-injection/)에 대해 학습해보시면 좋을 것 같습니다 ㅎㅎ

OikimiO replied:
장점
1. 코드가 간결
2. 본인이 아닌 DI Container가 의존성을 관리해 객체간의 결합도가 낮아짐 

단점
1. 무분별한 필드 주입으로 단일 책임 원칙이 깨질 수 있음
2. DI Container가 의존성을 관리해, DI Container가 아닌 환경에선 사용하기 어려움
3. field injection은 Reflection으로 주입받아 불변성을 보장할 수 없음
4. RunTime 환경에서 순환 의존이 발생하는 것을 알 수 있다는 단점이 있음

결론: Spring 4.x 이후부터는 Constructor Injection을 사용하기를 권장

xrabcde commented on src/main/java/roomescape/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,78 @@
+
+import java.util.List;
+
+@Controller
+public class AdminController {
```
한 컨트롤러에서 너무 많은 일을 해주고 있어요 😢
공통점이 있는 api끼리 컨트롤러를 분리해보는 것은 어떨까요?

OikimiO replied:
다음과 같이 구성 하려는데, 혹시 괜찮을까요? ㅎ

AS_IS > TO_BE
1. AdminController > ReservationController, ReservationTimeController

xrabcde commented on src/main/java/roomescape/admin/dto/ReadReservationResponse.java:
```diff
@@ -0,0 +1,30 @@
+            reservation.getTime());
+    }
+
+    public static ReadReservationResponse entityToDTO(Reservation reservation){
+        return new ReadReservationResponse(reservation);
+    }
```
정적 팩토리 메서드를 사용해주셨군요 ! 
[정적 팩토리 메서드 네이밍 컨벤션](https://inpa.tistory.com/entry/GOF-%F0%9F%92%A0-%EC%A0%95%EC%A0%81-%ED%8C%A9%ED%86%A0%EB%A6%AC-%EB%A9%94%EC%84%9C%EB%93%9C-%EC%83%9D%EC%84%B1%EC%9E%90-%EB%8C%80%EC%8B%A0-%EC%82%AC%EC%9A%A9%ED%95%98%EC%9E%90#%EC%A0%95%EC%A0%81_%ED%8C%A9%ED%86%A0%EB%A6%AC_%EB%A9%94%EC%84%9C%EB%93%9C_%EB%84%A4%EC%9D%B4%EB%B0%8D_%EA%B7%9C%EC%B9%99)에 대해서도 한 번 읽어보시면 좋을 것 같습니다 ☺️

OikimiO replied:
네이밍 컨벤션을 다음과 같이 변경하겠습니다 ㅎ

변경 원인
- 개발자가 임의로 만든 메서드명은 파악의 어려움

AS_IS > TO_BE
1. entityToDTO > from
2. entityToList > from
3. add - 제거
4. read > of

xrabcde commented on src/main/java/roomescape/admin/repository/AdminRepository.java:
```diff
@@ -0,0 +1,92 @@
+    public Long saveReservationTime(SaveReservationTimeRequest saveReservationTimeRequest) {
+        GeneratedKeyHolder keyHolder = new GeneratedKeyHolder();
+        this.jdbcTemplate.update(con -> saveReserveTimeStatement(con, saveReservationTimeRequest), keyHolder);
+        long id = keyHolder.getKey().longValue();
+
+        return id;
```
변수 할당하지 않고 즉시 return 해줄 수 있겠네용

OikimiO replied:
참고하겠습니다ㅎ

xrabcde commented on src/test/java/roomescape/admin/controller/AdminControllerTest.java:
```diff
@@ -0,0 +1,174 @@
+    SaveReservationRequest given_10_20분_예약 = new SaveReservationRequest("김민기","2024-01-10", 1L);
+
+
+    @Nested
+    @DisplayName("방탈출 예약은")
```
Nested 와 DisplayName 활용 👍👍

xrabcde commented on src/test/java/roomescape/admin/controller/AdminControllerTest.java:
```diff
@@ -0,0 +1,174 @@
+
+    @Nested
+    @DisplayName("방탈출 예약은")
+    class Describe_Resevation{
```
🤓

```suggestion
    class Describe_Reservation{
```

xrabcde commented on src/test/java/roomescape/admin/controller/AdminControllerTest.java:
```diff
@@ -0,0 +1,174 @@
+        class Context_Not_Use_For_Reservation{
+
+            @Test
+            @DisplayName("취소하실 수 있습니다.")
+            @Rollback
+            void delete_reservation_time(){
+                /*
+                save_reservation_time(given_예약시간_13_20);
+
```
이 테스트코드는 왜 주석처리 해두신건가요!! 사용하지 않는다면 삭제해주세요 ㅎㅎ

OikimiO replied:
넵넵 참고하겠습니다..! ㅎ

xrabcde commented on src/main/java/roomescape/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,78 @@
+        return BASE_PATH+"/reservation.html";
+    }
+
+    @GetMapping(BASE_PATH+"/time")
```
페이지 호출이 안되고 있는 것 같아요 ~ (호출 path를 확인해주세요 😉)

<img width="650" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/56033755/70396c34-f9ab-4933-a61d-30c8b2cf6d84">

xrabcde commented on src/main/java/roomescape/admin/entity/Reservation.java:
```diff
@@ -0,0 +1,54 @@
+        this.date = saveReservationRequest.date();
+    }
+
+    public static Reservation read(ResultSet rs){
+        return new Reservation(rs);
+    }
+
+    private Reservation(ResultSet rs) {
+        try {
+            this.id = rs.getLong("reservation_id");
+            this.name = rs.getString("reservation_name");
+            this.date = rs.getString("reservation_date");
+            this.time = ReservationTime.read(rs);
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
```
Reservation 의 관점에서 read 는 어떤 역할인건가요? 
ResultSet 을 Reservation 으로 변환하는 건 누구의 책임일까요? 
`조회` 라는 행위에 대해 더 고민해보시면 좋을 것 같습니다 !

OikimiO replied:
데이터 액세스 계층의 역할
   1. DB 접근에 대한 추상화
   2. CRUD관련 메서드 제공
   3. DB의 데이터를 애플리케이션의 객체로 매핑
   4. 트랜잭션 관리

도메인 계층의 역할
   1. 비지니스 규칙과 로직을 표현
   2. 도메인 모델간의 엔티티 관계 표현
   3. 시스템의 주요 구조와 설계를 결정하는 역할

Reservation의 관점에서의 read

- 도메인 계층에서 데이터 매핑을 하고 있음
- 이는 도메인 계층과 액세스 계층의 결합도를 높이게 됨

ResultSet을 Reservation으로 변환하는 책임은?
 - 데이터 매핑을 담당하는 Data Access 계층의 책임
 
해당 내용을 토대로 코드를 수정해보도록 하겠습니다. 
예리한 질문 감사합니다 ㅎ

xrabcde commented on src/test/java/roomescape/admin/controller/AdminControllerTest.java:
```diff
@@ -0,0 +1,174 @@
+            List<Object> reservations = objectMapper.readValue(jsonResponse, new TypeReference<List<Object>>() {});
+
+            assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+            assertThat(reservations.size()).isGreaterThan(0);
```
`isPositive()` 를 사용할 수 있을 것 같아요 

```suggestion
            assertThat(reservations.size()).isPositive();
```

xrabcde commented on src/test/java/roomescape/admin/controller/AdminControllerTest.java:
```diff
@@ -0,0 +1,174 @@
+            List<Object> reservations = objectMapper.readValue(jsonResponse, new TypeReference<List<Object>>() {});
+
+            assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+            assertThat(reservations.size()).isEqualTo(2);
```
`hasSize()` 를 사용할 수 있을 것 같아요 ~ 

```suggestion
            assertThat(reservations).hasSize(2);
```

OikimiO replied:
isEqualTo를 사용하지 않고 hasSize를 사용하는게 더 명확하고 간단하네요..!
참고하겠습니다ㅎ

xrabcde reviewed (changes requested):
안녕하세요 은호님 ~ 첫 번째 미션 리뷰를 맡은 조아라입니다 ☺️
요구사항에 맞게 잘 구현해주셨네요 ㅎㅎ
몇 가지 코멘트 남겨드렸는데 확인부탁드립니다 !
어렵거나 고민되는 점이 있다면 편하게 질문 남겨주세요 ~

OikimiO committed:
refactor: AdminController 기존에 있던 reservation, reservation time 메서드 제거

OikimiO committed:
refactor: reservation, reservationTime 관련 Controller와 Service추가

- Autowired 제거 후 생성자 주입으로 변경
- 기존 Admin에서 구현했던 reservation, reservationTime 메서드 분리

OikimiO committed:
refactor: entity read, add 메서드 제거 후 of 추가 및 데이터 매핑 메서드 제거

OikimiO committed:
refactor: dto 메서드 변경 및 ReservationService, ReservationTimeService 구현

- entityToDto > from, entityToList > from으로 메서드명 변경
- ReservationService, ReservationTimeService 의존성 주입하는 것으로 변경
- AdminService 제거

OikimiO committed:
refactor: ReservationRepository, ReservationTimeRepository 구현

- 생성자 주입을 통한 의존성 관리
- 데이터 매핑 작업을 위한 메서드 추가
- AdminRepository 제거

OikimiO committed:
refactor: 개선할 테스트 기능 변경 및 사용되지 않는 메서드 제거

xrabcde commented on src/main/java/roomescape/admin/controller/ReservationController.java:
```diff
@@ -0,0 +1,46 @@
+        this.reservationService = reservationService;
+    }
+
+    @GetMapping("/reservation")
+    public String reservation(){
+        return BASE_PATH+"/reservation.html";
+    }
```
화면 응답용 컨트롤러와 api 응답용 컨트롤러를 나눠보는 건 어떤가요?

xrabcde commented on src/main/java/roomescape/admin/controller/ReservationController.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @GetMapping("/reservations")
+    public ResponseEntity<List<ReadReservationResponse>> readReservation(){
+        List<ReadReservationResponse> response = this.reservationService.readReservation();
```
`readReservation()` 은 단건 조회와 헷갈릴 수 있어서 (이번 요구사항에는 없지만)
목록 조회라는 내용이 더 드러나는 네이밍을 지어줘도 좋을 것 같아요 ㅎㅎ

OikimiO replied:
쿼리 메서드 명명 규칙을 참고하여 구성을 변경했습니다. 
조회를 변경하면서 삭제/저장 메서드도 변경했습니다 ㅎ

AS_IS > TO_BE
 - DTO read~: find~
 - <List> read~: findAll
 - DTO save ~: save
 - void delet ~: delete

xrabcde commented on src/main/java/roomescape/admin/dto/ReadReservationResponse.java:
```diff
@@ -18,11 +18,11 @@ private ReadReservationResponse(Reservation reservation) {
     }
 
-    public static ReadReservationResponse entityToDTO(Reservation reservation){
+    public static ReadReservationResponse from(Reservation reservation){
```
👍👍

xrabcde commented on src/main/java/roomescape/admin/repository/ReservationTimeRepository.java:
```diff
@@ -0,0 +1,62 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    public List<ReservationTime> readReservationTime() {
+        return this.jdbcTemplate.query(ReservationTimeQuery.READ_ALL, readReservationTimeRowMapper());
```
여기도 여러건을 조회한다는 내용이 이름에 들어갈 수 있을 것 같아요 ㅎㅎ

xrabcde commented on src/main/java/roomescape/admin/service/ReservationTimeService.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.util.List;
+
+@Service
+public class ReservationTimeService {
+
+    private final ReservationTimeRepository reservationTimeRepository;
```
서비스를 분리하니 훨씬 책임과 역할이 명확해진 것 같아요 👍

xrabcde commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,20 @@
+    date VARCHAR(255) NOT NULL,
+    time_id BIGINT,                           -- 컬럼 수정
+    PRIMARY KEY (id),
+    FOREIGN KEY (time_id) REFERENCES reservation_time (id) -- 외래키 추가
```
외래키를 사용해주셨군요 !
시간에 대한 예약이 존재하는 상황에서 시간을 삭제하려고 하면 어떻게 될까요? 🤔
2단계 예외처리 진행하시면서 같이 해결해볼 수 있을 것 같아요 ㅎㅎ

xrabcde commented on src/test/java/roomescape/admin/controller/AdminControllerTest.java:
```diff
@@ -0,0 +1,153 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+public class AdminControllerTest {
```
테스트 코드도 운영코드에 맞춰 클래스 분리가 필요하겠네요 ~!

OikimiO replied:
2단계에선 참고하여 진행하겠습니다
감사합니다 ㅎ

xrabcde reviewed (approved):
안녕하세요 은호님 ~ 피드백 내용을 잘 반영해주셨네요 ㅎㅎ
이어서 2단계 진행해주시면 되겠습니다 ~!
